pipeline/trusted-resources/pkg/webhook/validation_admit.go

```diff
@@ -0,0 +1,405 @@
+		if tr.Spec.TaskSpec != nil {
+			logger.Info("Verifying TaskSpec")
+			if err := verifyTaskSpec(ctx, tr, tr.Spec.TaskSpec); err != nil {
+				return err
+			}
+			return nil
+		}
+
+		if tr.Spec.TaskRef != nil {
+			if tr.Spec.TaskRef.Bundle != "" {
+				logger.Info("Verifying OCI Bundle")
+				if err := ac.verifyTaskOCIBundle(ctx, tr); err != nil {
+					return err
+				}
+				return nil
+			}
+			task, err := ac.tektonClient.TektonV1beta1().Tasks(tr.Namespace).Get(ctx, tr.Spec.TaskRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if task.Name != "" {
+				logger.Info("Verifying TaskRef")
+				if err := verifyTaskSpec(ctx, tr, &task.Spec); err != nil {
+					return err
+				}
+				return nil
+			}
+		}
+		// TODO: Finish this in next PR
+		//case pipeline.PipelineRunControllerName:
+		//case pipeline.RunControllerName:
+
+	}
+
+	return nil
+}
+
+func verifyTaskSpec(ctx context.Context, tr *v1beta1.TaskRun, taskspec *v1beta1.TaskSpec) error {
+	ts, err := json.Marshal(taskspec)
+	if err != nil {
+		return err
+	}
+
+	// get checksum
+	checksum, err := getChecksum(ts)
+	if err != nil {
+		return err
+	}
+
+	rawPayload, err := json.Marshal(checksum)
+	if err != nil {
+		return err
+	}
+
+	if err := verifyTaskRunRawPayload(ctx, tr, rawPayload); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ac *reconciler) verifyTaskOCIBundle(ctx context.Context, tr *v1beta1.TaskRun) error {
+	serviceAccountName := os.Getenv("WEBHOOK_SERVICEACCOUNT_NAME")
+	if serviceAccountName == "" {
+		serviceAccountName = "tekton-pipelines-webhook"
+	}
+
+	kc, err := k8schain.New(ctx, ac.client, k8schain.Options{
+		Namespace:          system.Namespace(),
+		ServiceAccountName: serviceAccountName,
+	})
+	if err != nil {
+		return err
+	}
+
+	digest, err := getDigest(strings.TrimSuffix(strings.ToLower(string(tr.Spec.TaskRef.Kind)), "s"), tr.Spec.TaskRef.Bundle, kc)
+	if err != nil {
+		return err
+	}
+
+	rawPayload, err := json.Marshal(digest)
+	if err != nil {
+		return err
+	}
+
+	if err := verifyTaskRunRawPayload(ctx, tr, rawPayload); err != nil {
+		return err
+	}
+	return nil
+}
+
+func verifyTaskRunRawPayload(ctx context.Context, tr *v1beta1.TaskRun, rawPayload []byte) error {
+	if tr.ObjectMeta.Annotations != nil && tr.ObjectMeta.Annotations[signatureSource] != "" {
+		signatureEncoding := tr.ObjectMeta.Annotations[signatureSource]
+		signatureDecoding, err := base64.StdEncoding.DecodeString(signatureEncoding)
+		if err != nil {
+			return err
+		}
+		var publicKeyPath string
+		if tr.ObjectMeta.Annotations != nil && tr.ObjectMeta.Annotations[publicKey] != "" {
+			// Fetch public key from annotation
+			publicKeyPath = tr.ObjectMeta.Annotations[publicKey]
+		} else {
+			// Fetch public key from mounted path
+			publicKeyPath = SecretPath
+		}
+		// Verify
+		if err := verifySignature(ctx, signatureDecoding, rawPayload, publicKeyPath); err != nil {
+			return err
+		}
+
+	}else{
+		return fmt.Errorf("Wrong Annotations.")
```

Comment:
nit: add more details about what annotations were expected (e.g. if you were a user and all you saw this error - what details would be helpful for you to fix it?)